### ISSUE TITLE
Blocks: Inner block validation on the server side

### ISSUE THREAD

**Issue by tyxla** [WordPress/gutenberg#51087]
## What problem does this address?
Currently, we're not doing any server-side validation for inner blocks based on their `parent` or `ancestor` field. Because of that, if a block gets incorrectly inserted as an inner block of an unsupported block (one that's not within the `parent` or `ancestor` fields), that can lead to unexpected results.

We recently discovered that by noticing that inside a `core/navigation` block was inserted a `wp:pattern`, and that pattern included a navigation block inside a Group block, and rendering them led to an infinite loop on the backend. 

Since it's not clear to me how a pattern that is wrapped with a Group block can be inserted inside a navigation block, I've opened another issue for this, in hopes of finding a way to reproduce that: #51086.

## What is your proposed solution?

One potential way to address this could be filtering all inner blocks, and removing them if they're not supported when adding them as inner blocks in `WP_Block_List` or in `WP_Block`.
